html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm

```diff
@@ -30,8 +30,22 @@ has '+sources' => (isa => 'ArrayRef['.join('|', @{sources_classes()}).']');
 
 has '+multi_source_object_classes' => (default => sub{sources_classes()});
 
+has 'challenge_template' => (is => 'ro', default => sub { "challenge.html" } );
+
+has 'challenge_data' => (is => 'rw', builder => '_build_challenge_data', lazy => 1, trigger => \&_trigger_challenge_data);
+
+sub _build_challenge_data {
+    my ($self) = @_;
+    return $self->app->session->{challenge_data};
```

Comment:
Would be better to use the module session IMO. This way, if there is a chain of logins, it will not need to be cleared.

`$self->session->{challenge_data}`